pkg/config/new.go

```diff
@@ -0,0 +1,246 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/BurntSushi/toml"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	cachedConfigError error
+	cachedConfigMutex sync.Mutex
+	cachedConfig      *Config
+)
+
+const (
+	// FIXME: update code base and tests to use the two constants below.
+	containersConfEnv         = "CONTAINERS_CONF"
+	containersConfOverrideEnv = containersConfEnv + "_OVERRIDE"
+)
+
+// Options to use when loading a Config via New().
+type Options struct {
+	// Attempt to load the following config modules.
+	Modules []string
+
+	// Set the loaded config as the default one which can later on be
+	// accessed via Default().
+	SetDefault bool
+
+	// Additional configs to load.  An internal only field to make the
+	// behavior observable and testable in unit tests.
+	additionalConfigs []string
+}
+
+// New returns a Config as described in the containers.conf(5) man page.
+func New(options *Options) (*Config, error) {
+	if options == nil {
+		options = &Options{}
+	} else if options.SetDefault {
+		cachedConfigMutex.Lock()
+		defer cachedConfigMutex.Unlock()
+	}
+	return newLocked(options)
+}
+
+// Default returns the default container config.  If no default config has been
+// set yet, a new config will be loaded by New() and set as the default one.
+// All callers are expected to use the returned Config read only.  Changing
+// data may impact other call sites.
+func Default() (*Config, error) {
+	if cachedConfig != nil || cachedConfigError != nil {
```

Comment:
Double checked locking (on its own) is not safe with Golang's memory model. While you may observe the write to cachedConfig, there is no happens-before guarantee on the preceding writes for the config structure as part of initialization (the pointer write may be reordered (speculative execution, compiler reordering etc) or coherence not yet achieved leading to a program to observe an uninitialized / incomplete config). See https://go.dev/ref/mem. Also https://gee.cs.oswego.edu/dl/jmm/cookbook.html is pretty dated but still nice overview/example of how compilers map a model to hardware. 

I'd recommend using either sync.Once or an atomic Load Store for the pointer in combo with the DCL as it is (what Once does under the hood) . The go memory model does have a happens-before established at the location of an atomic store, but without the lock queuing or CAS overhead.

Reply:
Thanks a lot, @n1hility !  Will fix accordingly

Reply:
@vrothberg +1, only reason i suggested sync.Once was if you were trying to avoid the lock penalty. putting the Lock/Unlock in Do would give you that same DCL lock for just init behavior. I agree the opto isn't critical in this case.

Reply:
Thanks a lot for checking!  I did consider a sync.Once but with this PR `New()` may override the "default" config which would have broken the contract Once would guarantee otherwise.